### PR TITLE
Add daily CleanupJob to remove stale DB entries

### LINKED ISSUE #44: Clear jobs from the db
So as not to clog up the db there should be a way to clear jobs from the db after a set interval, or after x completions. Use case being I have a job that runs every minute, over the course of a day thats 1440 new records.

### PR DESCRIPTION
**Pull request by andrewandante**
Fixes #44 

Introduces a CleanupJob that is off by default to ensure an upgrade wouldn't start destroying DB entries unexpectedly.

Cleans out anything in the DB with JobStatus "Complete" or "Broken" more than 30 days old in its default configuration. Can be tweaked to save n records, then clean out anything matching a valid status above that.

Tweaked the end of `runJob()` to ensure there is one CleanupJob in the queue, and if not, schedule on in for a day from then. Should mean that it runs once a day.

### CODE REVIEW

File: README.md
Comment (dhensby) [inline README.md]: > it is the total number of records to keep.

Is it not the *minimum* number to keep?

File: code/jobs/CleanupJob.php
Comment (dhensby) [inline code/jobs/CleanupJob.php]: These should be `@var` not `@return`

**Review by dhensby — CHANGES_REQUESTED** [review]
Looking good - just a few minor docs issues

File: README.md
Comment (andrewandante) [inline README.md]: True. Will update.

File: code/jobs/CleanupJob.php
Comment (andrewandante) [inline code/jobs/CleanupJob.php]: Whoops.

**Review by dhensby — APPROVED** [review]

**Comment by Firesphere** [comment]
Have you thought of adding tests?

**Comment by andrewandante** [comment]
Do you mean in addition to `tests/CleanupJobTest.php`? Were you envisaging others?

**Comment by Firesphere** [comment]
I'm blind. Or you secretly force-pushed with an adjusted date. I'm guessing the first.

**Comment by andrewandante** [comment]
@nyeholt any movement or feedback on the above?

File: code/services/QueuedJobService.php
Comment (nyeholt) [inline code/services/QueuedJobService.php]: Could the job itself be responsible for re-adding itself to the queue? The initial creation can be done directly from the UI - my thinking being that if you want to _not_ have the cleanup running, you're needing to make server level configuration changes rather than just deleting the pending job from the UI.

Comment (andrewandante) [inline code/services/QueuedJobService.php]: It could - I did consider that behaviour. As it is already disabled by default, I'm not sure which I prefer. I think it ultimately comes down to who you want controlling whether or not the cleanup job runs - the CMS admin, or the dev. My gut says the dev, and that it should be controlled by config, but I guess ultimately it is up to you as module maintainer. Happy to change it over to be a self-restarting job if that's what you think is best.

Comment (nyeholt) [inline code/services/QueuedJobService.php]: If your deployment process is something that takes 10s of minutes across multiple distributed backend nodes, that requires a convoluted approval process to perform, I'd much prefer being able to simply remove it from the UI and not have it restart rather than go through that if I need to stop it from running.

Comment (andrewandante) [inline code/services/QueuedJobService.php]: Point taken. Moved the regeneration of the job to the job itself!

**Comment by dhensby** [comment]
@nyeholt ping

**Comment by nyeholt** [comment]
Sorry, didn't get the followup commit notification - all good!

**Pull request merged**
